### ISSUE TITLE
JS Heap out of Memory when Aligning (with isolate set to true)

### ISSUE THREAD

**Issue by LordMartron94** [echogarden-project/echogarden#69]
Code:
```py
def align(self, video_path: Path, transcript_path: Path):
		subtitle_path = video_path.with_suffix(".srt")
		command = [
			"align",
			f"\"{video_path}\"",
			f"\"{transcript_path}\"",
			f"\"{subtitle_path}\"",
			"--engine=whisper",
			"--language=el",
			"--subtitles.mode=sentence",
			# f"--recognition.whisper.model={MODEL_USED}",
			f"--whisper.model={MODEL_USED}",
			"--isolate=true",
			"--sourceSeparation.mdxNet.model=UVR_MDXNET_KARA"
		]

		command_helper: CommandHelper = CommandHelper(HOORN_LOGGER)
		command_helper.execute_command_v2("echogarden", command, shell=True)
```

MODEL_USED is "small"... I changeed the sourceseparation model only to test if the issue would go away there.

Output:
Echogarden v1.5.0

Transcode with command-line ffmpeg.. 719.6ms
Convert wave buffer to raw audio.. 115.9ms

Prepare for source separation.. 
Downloading mdxnet-UVR_MDXNET_KARA-20240330.. (27.52MB): 0% 20% 60% 100% (2.02s, 13.63MB/s)
2219.1ms
Compute STFT of full waveform.. Uncaught exception:
 Error [ERR_WORKER_OUT_OF_MEMORY]: Worker terminated due to reaching memory limit: JS heap out of memory

**Comment by LordMartron94** [comment]
The error is weird because I have 64GB worth of memory and it was utilized in total only about halfway.

**Comment by rotemdan** [comment]
The error happens when it computes the STFT (Short Term Fourier Transform) of the audio file. 

It's possible the input waveform is very large and this may require more memory than the Node.js runtime can handle.

The reason this could happen is because:

```ts
const fftFramesLeft = await stftr(rawAudio.audioChannels[0], fftSize, fftWindowSize, fftHopSize, 'hann')
const fftFramesRight = await stftr(rawAudio.audioChannels[1], fftSize, fftWindowSize, fftHopSize, 'hann')
```

Computes the STFT of the entire waveform (each channel separately), all at once, meaning that the entire result is stored in memory.

A possible future fix would perform this processing in a more streaming way.

In any case, the implementation of MDXNet is still a prototype, and is currently much slower than others, like the ones in [Ultimate Vocal Remover GUI](https://github.com/Anjok07/ultimatevocalremovergui) and [Python Audio Separator](https://github.com/nomadkaraoke/python-audio-separator) (Python based tools), mostly due to GPU acceleration for STFT computations, which I plan to add as well in the future.

@beskrovnykh The error you are describing sounds unrelated to this and could be more closely related to #64, which is fixed in `1.6.0` but I'm not exactly sure based on the information I have.

**Comment by rotemdan** [comment]
On `v1.7.0` I've made a partial rewrite of the voice isolation operation (MDX-NET implementation), which is much faster (especially with GPU acceleration, which now gets to 20x real-time on my machine) and fixes the memory issue you describe here. (also 3 new higher-quality models, which you can read more about on the [release notes](https://github.com/echogarden-project/echogarden/releases/tag/v1.7.0)).

The memory issue was caused by several factors:

* It precomputed all STFT frames in advance. Now it uses an incremental frame generator and efficient windowed list structure to allow for overlapped processing.
* It continuously allocated a very large amount of small object structures for complex numbers (like `{ real: number, imaginary: number }`) that produced very high pressure on the garbage collector, which the V8 runtime couldn't handle efficiently. The new approach uses flattened typed arrays (real and complex parts are flattened sequentially) that don't have that problem. It turned out this was actually the stronger cause of the issue, and also the cause of the previous instability with DirectML enabled.

The longest I've tested it so far was a 1 hour 15 minute audio input, which took 3 minutes to process (GPU enabled) and reached a maximum of 3GB of memory.

**Comment by LordMartron94** [comment]
> On `v1.7.0` (just released) I've made a partial rewrite of the voice isolation operation (MDX-NET implementation), which is now much faster (especially with GPU acceleration, which now gets to 20x real-time on my machine) and fixes the memory issue you describe here. (also added 3 new higher-quality models, which you can read more about on the [release notes](https://github.com/echogarden-project/echogarden/releases/tag/v1.7.0)).
> 
> The memory issue was caused by several factors:
> 
>     * It precomputed all STFT frames in advance. Now it uses an incremental frame generator and efficient windowed list structure to allow for overlapped processing.
> 
>     * It continuously allocated a very large amount of small object structures for complex numbers (like `{ real: number, imaginary: number }`) that produced very high pressure on the garbage collector, which the V8 runtime couldn't handle efficiently. The new approach uses flattened typed arrays (real and complex parts are flattened sequentially) that don't have that problem. It turned out this was actually the stronger cause of the issue, and also the cause of the previous instability with DirectML enabled.
> 
> 
> The longest I've tested it so far was a 1 hour 15 minute audio input, which took 3 minutes to process (GPU enabled) and reached a maximum of 3GB of memory.

Holy shit, that actually sounds amazing. I'll test it out later when I have another file I need to translate. Thanks for your amazing work. :)

**Comment by rotemdan** [comment]
Thanks. On `1.8.x` the speed improved to about 30x real-time on the default model on my RTX 2060 / 13th-gen Core i3 due to switching to use a new WASM build of `PFFFT` I made, with SIMD enabled. On the higher quality models like `Kim Vocal 2` I measured about 17x real-time. This on Windows with `mdxNet.provider=dml` (DirectML), which is now selected by default, with CPU fallback. On Linux I tested it with `mdxNet.provider=cuda` to have same or slightly faster performance.

For CPU-only processing (`mdxNet.provider=cpu`) I'm seeing about 4.7x real-time for the default model and 1.625x real-time for the higher quality models on 13th-gen Core i3 (4 cores, 8 threads).

It's actually now as fast, or a bit faster than [Ultimate Vocal Remover GUI](https://github.com/Anjok07/ultimatevocalremovergui) and [Python Audio Separator](https://github.com/nomadkaraoke/python-audio-separator), but they provide other model architectures like Demucs and VR Architecture, many more models and other features like instrument separation, model ensembling, more customization, etc.
